File: dsa/patterns/21_greedy/python/job_sequencing_problem.py
# Problem: Job Sequencing Problem
# Idea: Sort jobs by profit in descending order, then for each job find the latest free slot before its deadline
# Delay the jobs as late as possible to leave first slots for other jobs
# Time: O(n * max_deadline)
# Space: O(max_deadline)

def job_sequencing(jobs):
    # jobs = [(jobId, deadline, profit), ...]
    
    # sort jobs by profit descending
    jobs.sort(key=lambda x: x[2], reverse=True)

    # find maximum deadline
    max_deadline = 0
    for _, d, _ in jobs:
        max_deadline = max(max_deadline, d)

    # slot array to track occupied time slots
    slots = [-1] * (max_deadline + 1)

    job_count = 0
    total_profit = 0

    for job_id, deadline, profit in jobs:
        # try to place job at latest possible slot
        for t in range(deadline, 0, -1):
            if slots[t] == -1:
                slots[t] = job_id
                job_count += 1
                total_profit += profit
                break

    return job_count, total_profit


# -------- Test --------
jobs = [
    (1, 4, 20),
    (2, 1, 10),
    (3, 2, 40),
    (4, 2, 30)
]

count, profit = job_sequencing(jobs)
print(count, profit)   # Expected: 3 90

File: dsa/patterns/21_greedy/java/JobSequencingProblem.java
import java.io.IOException;
import java.util.Arrays;

class JobSequencingProblem {
	public static void main(String[] args) {

		// size of array
		Job[] arr = new Job[4];
		arr[0] = new Job(1, 4, 20);
		arr[1] = new Job(2, 1, 10);
		arr[2] = new Job(3, 2, 40);
		arr[3] = new Job(4, 2, 30);

		JobSequencingProblem obJ = new JobSequencingProblem();

		// function call
		int[] res = obJ.schedule(arr, 4);
		System.out.println(res[0] + " " + res[1]); // count of jobs and max profit made

	}

	// Idea: Sort jobs by profit in descending order, then for each job find the latest free slot before its deadline
	// Delay the jobs as late as possible to leave first slots for other jobs
	// Time: O(n*max_deadline)
	public int[] schedule(Job arr[], int n) {
		Arrays.sort(arr, (j1, j2) -> (j2.profit - j1.profit));

		int maxi = 0;
		for (int i = 0; i < n; i++)
			maxi = Math.max(maxi, arr[i].deadline);

		int hash[] = new int[maxi + 1];
		Arrays.fill(hash, -1);

		int jobsCnt = 0, jobsProfit = 0;
		for (int i = 0; i < n; i++) {
			for (int j = arr[i].deadline; j >= 0; j--) {
				if (hash[j] == -1) { // Free slot found
					hash[j] = i;
					jobsCnt++;
					jobsProfit += arr[i].profit;
					break;
				}
			}
		}

		return new int[] { jobsCnt, jobsProfit };
	}
}

class Job {
	int id, profit, deadline;

	Job(int id, int deadline, int profit) {
		this.id = id;
		this.deadline = deadline;
		this.profit = profit;
	}
}
